api/envoy/extensions/filters/http/jwt_authn/v3/config.proto

```diff
@@ -455,12 +455,22 @@ message RemoteJwks {
   config.core.v3.RetryPolicy retry_policy = 4;
 
   // Refetch JWKS if extracted JWT has no KID or a KID that does not match any cached JWKS's KID.
-  // :ref:`Asynchronous JWKS fetching <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.RemoteJwks.async_fetch>` is independent to
-  // this refetching process.
   //
   //
-  // In envoy, new JWKS is fetched on demand and the cache is isolated to the fetched worker thread. If none of the cached JWKS entries contain a KID,
-  // each incoming request will initiate a new JWKS fetch.
+  // In envoy, if
+  // :ref:`async JWKS fetching <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.RemoteJwks.async_fetch>`
```

Comment:
Is it possible for multiple threads to each spawn a fetch at the same time?

Is it possible for the main thread to have an async fetch already in progress when we trigger a new fetch in a different thread?

I think we want the answer to both of those to be no, but the wording here doesn't make it clear.  Maybe this should say something like the following:

"""
If async JWKS fetching is enabled along with this field, then KID mismatch will trigger a new async fetch after appropriate backoff delay.
"""

Reply:
The answer to both questions is no, but as for the wording "then KID mismatch will trigger a new async fetch after appropriate backoff delay." , not sure if we are talking about backoff in the same context here. Are you referring to the async fetch timer being reset to the default value if we retrieve a new JWKS due to KID mismatch?

Backoff only occurs for two scenarios (Async is unrelated):
Successful KID fetch when KID mismatch happens due to no KID in JWT. 
Failed KID fetch/Verification on any KID mismatch.

Reply:
My concern is with the "KID mismatch will trigger a fetch in the worker thread" part.  What happens if the main thread already has a fetch in flight at that moment?  What happens if multiple worker threads see the KID mismatch at the same moment?  Also, it seems better to not specify what thread the fetch will actually happen in, since that's an implementation detail.

That's why I suggested just saying something like "KID mismatch will trigger a new async fetch" -- the implication there is that there's only ever one fetch in flight at once, and its result is visible to all threads.  I think that's actually the important property; things like which thread the fetch actually happens in seems like an implementation detail that is likely to be different in different data planes.

With regard to the backoff delay, consider the following sequence of events:
1. The main thread does an async fetch, gets a result, and makes the result available to the worker threads.
2. Worker thread 1 gets a request and grabs the latest result from the main thread.  It sees a KID mismatch.

At this point, I think we should impose backoff before we re-fetch, since the main thread just finished a fetch moments ago.  This probably means that we should fail the request, since we can't do a new fetch immediately.  Are we doing that correctly?

Reply:
Yeah, we are. And I misunderstood what you meant initially thinking it meant there was always going to be a backoff with async, irrespective of circumstances. Thanks for clarifying.

I'll update the docs to what you suggested. Will also add tests since we seem to be in agreement about the overall flow.

Reply:
I've updated it. This is ready for review.